### PR TITLE
Add gRPC implementation to GCS API

### LINKED ISSUE #3280: Make GCS API available via gRPC
#### Summary
The Gateway Configuration Server currently does not use gRPC so it's API is plain HTML. In the JS SDK we somewhat rely on functionality being available as RPC so adding GCS functionality requires us to break out of our usual implementation.

#### Why do we need this?
- Retain implementation consistency in the JS SDK
- Avoid harcoded API routes and arbitrary implementations

#### What is already there? What do you see now?
gRPC API of all other stack components

#### What is missing? What do you want to see?
gRPC API for the GCS

#### How do you propose to implement this?
Likely recreate the implementation for the other stack components.

#### How do you propose to test this?
Unit tests.

#### Can you do this yourself and submit a Pull Request?
No, @KrishnaIyer @adriansmares please reassign as appropriate.

### PR DESCRIPTION
**Pull request by nicholaspcr**
#### Summary

Closes #3280

#### Changes

- Add proto definition of GCS service
- Add grpc implementation of GCS service
- Add gRPC tests

#### Testing

Unit tests and manually testing by downloading the global config in the gateway page of console.

##### Regressions

The gRPC service is adjacent to the http routes that already exists, meaning that the route generated by the `grpc-gateway` is distinct from the previous routes. 

#### Notes for Reviewers

After the discussion in this PR thread, it was decided to create the gRPC http routes alongside with the previous http routes. This avoid the API breaking, in practical terms the http routes become something similar to: 
- (old_http) : `http://localhost:1885/api/v3/gcs/gateways/{gateway_id}/semtechudp/global_conf.json`
- (grcp_http): `http://localhost:1885/api/v3/gcs/gateways/configuration/{gateway_id}/{format}/{filename}`

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/gateway_configuration.proto
Comment (htdvisser) [inline api/gateway_configuration.proto]: This probably needs some validation.

**Review by htdvisser — COMMENTED** [review]
I wasn't involved in the original discussion about building this feature, so this may be a bit late, but still. I don't think that "in the JS SDK we expect everything to be gRPC" is a good argument for forcing something that should be a simple file download into a gRPC API. Are we sure we really need to do this here? Why can't the SDK just support the file downloads that we already have?

File: api/gateway_configuration.proto
Comment (nicholaspcr) [reply api/gateway_configuration.proto]: I'm not sure what should be allowed in the format so I opted to just leave lowercase letters with hyphens. Let me know what you think.

File: CHANGELOG.md
Comment (johanstokking) [inline CHANGELOG.md]:
```suggestion

- gRPC service to Gateway Configuration Server so that gateway configurations can be obtained via gRPC requests.
- Retain at most 10 recent session keys in the Join Server. This avoids a slowly growing number of session keys in the Join Server's database.
```

File: api/gateway_configuration.proto
Comment (johanstokking) [inline api/gateway_configuration.proto]:
```suggestion

```

File: api/gateway_configuration.proto
Comment (johanstokking) [inline api/gateway_configuration.proto]: Maybe use the ID regex we use everywhere?
For sure include a length check

File: pkg/gatewayconfigurationserver/grpc.go
Comment (johanstokking) [inline pkg/gatewayconfigurationserver/grpc.go]:
```suggestion
// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
```

**Review by johanstokking — COMMENTED** [review]
Please rebase and `make init` because `data/lorawan-devices` diff doesn't belong here

File: pkg/gatewayconfigurationserver/grpc.go
Comment (johanstokking) [reply pkg/gatewayconfigurationserver/grpc.go]: Also at least one other place

File: api/gateway_configuration.proto
Comment (nicholaspcr) [reply api/gateway_configuration.proto]: was going to do that initially but wasn't sure if format should include numbers or not. Since its fine will use that same regex as the others

File: CHANGELOG.md
Comment (johanstokking) [reply CHANGELOG.md]: This isn't addressed

File: CHANGELOG.md
Comment (johanstokking) [inline CHANGELOG.md]: See my previous review comment

**Review by johanstokking — APPROVED** [review]

**Review by htdvisser — APPROVED** [review]
LGTM, but please also wait for review from @KrishnaIyer as main GCS code-owner

File: pkg/gatewayconfigurationserver/grpc.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/grpc.go]: You can add the format as an attribute to the error (ex: `{format}`).

File: pkg/gatewayconfigurationserver/server.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/server.go]: The service is `GatewayConfigurationService` so shouldn't that be registered in the hook?

File: pkg/gatewayconfigurationserver/grpc.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/grpc.go]: And use this format here as an attribute.

File: pkg/gatewayconfigurationserver/http.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/http.go:38]: Aren't we replacing these routes with the gRPC service? 

The gRPC gateway paths are defined in your new proto as 
```
"/gcs/gateways/{gateway_ids.gateway_id}/configuration/{format}"
```

If you want a Semtech UDP file, you'd make a request like

1. Through gRPC gateway
```
/gcs/gateways/{gateway_ids.gateway_id}/configuration/semtechudp
```

2. Through the existing web route
```
/gcs/gateways/{gateway_id}/semtechudp/global_conf.json
```

These look like they do the same thing. 

If you add a `string filename` field to your `GetGatewayConfigurationRequest` message and remove the "configuration" word, then the gRPC request would become
```
/gcs/gateways/{gateway_ids.gateway_id}/{format}/{filename}
```

This is the same as what's duplicated in these web routes right? So why not make harmonize these and get rid of the web routes?

File: pkg/gatewayconfigurationserver/grpc.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/grpc.go:58]: I can't really think of a valid use case for disabling auth here. The configuration for a specific gateway is never public right?  We had this already in the codebase but I can't remember why. @htdvisser ?

**Review by KrishnaIyer — CHANGES_REQUESTED** [review]

File: pkg/gatewayconfigurationserver/http.go
Comment (nicholaspcr) [reply pkg/gatewayconfigurationserver/http.go:38]: I see it, the http and grpc routes could be merged. In addition of the `filename` field could open space for more configuration files without adding extra routes. 

Will work on that change.

Comment (nicholaspcr) [reply pkg/gatewayconfigurationserver/http.go:38]: It slipped by me when looking it at first  but there is a problem with this solution, it becomes more clear when looking at the previously existent paths:
- `/gcs/gateways/{gateway_id}/semtechudp/global_conf.json`
- `/gcs/gateways/{gateway_id}/kerlink-cpf/lorad/lorad.json`
- `/gcs/gateways/{gateway_id}/kerlink-cpf/lorafwd/lorafwd.toml`

The suggestion `/gcs/gateways/{gateway_ids.gateway_id}/{format}/{filename}` becomes something where `{format}` has to include something like `kerlink-cpf/loradfwd` which from my tests locally doesn't seem possible to do.

I'm a fan of making a gRPC gateway that replaces the the three previously defined http routes but I'm struggling to see a solution that doesn't change the api. So in my mind it might be more worth it to leave it as is.

File: pkg/gatewayconfigurationserver/server.go
Comment (nicholaspcr) [reply pkg/gatewayconfigurationserver/server.go]: Wasn't sure if you were referring to the namespace name or the `GCs` in the filter. 

Either way I changed the filter to refer to the `GatewayConfigurationService` instead of `GCs` and changed the namespace from `gatewayserver` to `gatewayconfigurationserver`.

File: pkg/gatewayconfigurationserver/http.go
Comment (nicholaspcr) [reply pkg/gatewayconfigurationserver/http.go:38]: @KrishnaIyer small reminder of this discussion

Comment (KrishnaIyer) [reply pkg/gatewayconfigurationserver/http.go:38]: Ok in that case can you add a `type` parameter? 
Then you can have `/gcs/gateways/{gateway_ids.gateway_id}/{format}/{type}/{filename}`.

For the Semtech UDP, since the type is not needed, you can add an [additional binding](https://github.com/TheThingsNetwork/lorawan-stack/blob/1c018c71892ef2c258b99fc751ca44efa1abb8a3/api/application_services.proto#L165) to that path without the type.

Comment (nicholaspcr) [reply pkg/gatewayconfigurationserver/http.go:38]: From our discussion and tests, the problem with replacing the existent routes with the ones that are generated by the gRPC-gateway is the fact that the response body changes in such a way that the json of the configuration stays as the base64 of the byte array. So when doing a request to the GCS gRPC service the response will alway be:
```json
{
   "contents": "<base64_byte_array>"
}
```

Since this changes the API, is best to keep the old implementation.

File: pkg/gatewayconfigurationserver/grpc.go
Comment (htdvisser) [reply pkg/gatewayconfigurationserver/grpc.go:58]: At some point in the past we allowed Basic Station gateways to connect without auth. We don't do that anymore on any of our production deployments, but it's not impossible that some private deployments still allow it.

**Review by htdvisser — APPROVED** [review]
Nothing to add

File: pkg/gatewayconfigurationserver/grpc.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/grpc.go]: Nit. Since the error type is already `InvalidArgument`, I'd recommend replacing the word `Invalid` with `Unsupported` or `NotSupported` in the variable and the error name.

File: pkg/gatewayconfigurationserver/grpc.go
Comment (KrishnaIyer) [reply pkg/gatewayconfigurationserver/grpc.go:58]: Yes but these routes are not for LBS. These are those json/toml files for UDP gateways. So now anyone can just insert a gateway ID here and get the configuration for a particular gateway. 
Not that it's a huge security issue but I think we can avoid some confusion here.

File: pkg/gatewayconfigurationserver/grpc.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/grpc.go]: Nit: You can remove these extra new lines.

File: pkg/gatewayconfigurationserver/grpc.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/grpc.go]: I think it would be good to also log the type here.
```
configuration filename `{filename}` for type `{type}` is not supported
```

Also in some of the other log statements.

File: pkg/gatewayconfigurationserver/http.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/http.go:34]: Can you please add some notes here as to why these routes exists even though we have gRPC routes?

**Review by KrishnaIyer — COMMENTED** [review]
Looks really good. I only have tiny comments.

Have you tested that the old routes (currently used by the console) work with these changes?

File: pkg/gatewayconfigurationserver/http.go
Comment (nicholaspcr) [reply pkg/gatewayconfigurationserver/http.go:34]: Left in the godoc a bit of explanation of why this isn't the only route capable of return the gateway configuration and the reason for such occurrence. Let me know what you think

**Comment by nicholaspcr** [comment]
> Looks really good. I only have tiny comments.
> 
> Have you tested that the old routes (currently used by the console) work with these changes?

Just to be sure I tested again, the gateway configuration is returning the expected object in the old routes and in the new routes the response is the bytes of such object. Also checked if the `Download global_conf.json` button in the console returns what is expected, which it does.

File: pkg/gatewayconfigurationserver/http.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/http.go:35]: I don't think it's Go convention to leave an empty line but I might be wrong.

File: pkg/gatewayconfigurationserver/http.go
Comment (KrishnaIyer) [inline pkg/gatewayconfigurationserver/http.go]: I suggest the following
```
The gateway configuration value returned by the `grpc-gateway` routes are not content formatted, but a stream of bytes. This would be a breaking change for the consumers of this API and hence these routes are retained.
```

**Review by KrishnaIyer — APPROVED** [review]
LGTM

File: pkg/gatewayconfigurationserver/http.go
Comment (nicholaspcr) [reply pkg/gatewayconfigurationserver/http.go:35]: I believe it is part of the convention, specially since there are some sections of the standard library that utilizes that style of comment. For example the `Seek` method in the `os` pkg:
```go
// Seek sets the offset for the next Read or Write on file to offset, interpreted
// according to whence: 0 means relative to the origin of the file, 1 means
// relative to the current offset, and 2 means relative to the end.
// It returns the new offset and an error, if any.
// The behavior of Seek on a file opened with O_APPEND is not specified.
//
// If f is a directory, the behavior of Seek varies by operating
// system; you can seek to the beginning of the directory on Unix-like
// operating systems, but not on Windows.
func (f *File) Seek(offset int64, whence int) (ret int64, err error) {
	if err := f.checkValid("seek"); err != nil {
		return 0, err
	}
	r, e := f.seek(offset, whence)
	if e == nil && f.dirinfo != nil && r != 0 {
		e = syscall.EISDIR
	}
	if e != nil {
		return 0, f.wrapErr("seek", e)
	}
	return r, nil
}
```

I believe it's commonly used for package documentation, I opted to do that here to add a line just to separate the comment regarding the api in its own paragraph.

**Pull request merged**
